File: app/src/main/java/giuliolodi/gitnav/Adapters/EventAdapter.java
/*
 * Copyright 2017 GLodi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package giuliolodi.gitnav.Adapters;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.vstechlab.easyfonts.EasyFonts;

import org.eclipse.egit.github.core.event.CommitCommentPayload;
import org.eclipse.egit.github.core.event.DeletePayload;
import org.eclipse.egit.github.core.event.DownloadPayload;
import org.eclipse.egit.github.core.event.Event;
import org.eclipse.egit.github.core.event.FollowPayload;
import org.eclipse.egit.github.core.event.ForkPayload;
import org.eclipse.egit.github.core.event.GistPayload;
import org.eclipse.egit.github.core.event.GollumPayload;
import org.ocpsoft.prettytime.PrettyTime;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.hdodenhof.circleimageview.CircleImageView;
import giuliolodi.gitnav.GistActivity;
import giuliolodi.gitnav.R;
import giuliolodi.gitnav.RepoActivity;
import giuliolodi.gitnav.UserActivity;

public class EventAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private Context context;
    private List<Event> eventList;
    private PrettyTime p = new PrettyTime();

    public class EventHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.row_event_ll) LinearLayout linearLayout;
        @BindView(R.id.row_event_name) TextView name;
        @BindView(R.id.row_event_description) TextView description;
        @BindView(R.id.row_event_image) CircleImageView imageView;
        @BindView(R.id.row_event_date) TextView date;

        public EventHolder(View view) {
            super(view);

            ButterKnife.bind(this, view);

            name.setTypeface(EasyFonts.robotoRegular(context));
            description.setTypeface(EasyFonts.robotoRegular(context));
        }

    }

    public class EventLoadHolder extends RecyclerView.ViewHolder {

        public EventLoadHolder(View view) {
            super(view);
        }

    }

    @Override
    public int getItemViewType(final int position) {
        return eventList.get(position) != null ? 1 : 0;
    }

    public EventAdapter(Context context, final List<Event> eventList) {
        this.context = context;
        this.eventList = eventList;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder vh;
        if (viewType == 1) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_event, parent, false);
            vh = new EventHolder(itemView);
        } else {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_loading, parent, false);
            vh = new EventLoadHolder(itemView);
        }
        return vh;
    }

    private View.OnClickListener RepoListener (final String owner, final String repo) {
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                context.startActivity(new Intent(context, RepoActivity.class).putExtra("owner", owner).putExtra("name", repo));
                ((Activity) context).overridePendingTransition(0, 0);
            }
        };
    }

    private View.OnClickListener UserListener (final String user) {
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                context.startActivity(new Intent(context, UserActivity.class).putExtra("userS", user));
                ((Activity) context).overridePendingTransition(0, 0);
            }
        };
    }

    private View.OnClickListener GistListener (final String gistId) {
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                context.startActivity(new Intent(context, GistActivity.class).putExtra("GistId", gistId));
                ((Activity) context).overridePendingTransition(0, 0);
            }
        };
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {

        if (holder instanceof EventHolder) {

            final EventHolder h = ((EventHolder)holder);
            final Event e = eventList.get(position);
            h.name.setText(e.getActor().getLogin());
            Picasso.with(context).load(e.getActor().getAvatarUrl()).centerCrop().resize(75,75).into(h.imageView);
            h.date.setText(p.format(e.getCreatedAt()));
            h.name.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    context.startActivity(new Intent(context, UserActivity.class).putExtra("userS", e.getActor().getLogin()));
                    ((Activity) context).overridePendingTransition(0, 0);
                }
            });
            h.imageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    context.startActivity(new Intent(context, UserActivity.class).putExtra("userS", e.getActor().getLogin()));
                    ((Activity) context).overridePendingTransition(0, 0);
                }
            });

            String[] s = e.getRepo().getName().split("/");

            if (e.getPayload() != null) {

                switch (eventList.get(position).getType()) {

                    case "CommitCommentEvent":  // #1   NEED TEST
                        CommitCommentPayload commitCommentPayload = (CommitCommentPayload) e.getPayload();
                        h.description.setText(commitCommentPayload.getComment().getBody());
                        h.linearLayout.setOnClickListener(RepoListener(s[0], s[1]));
                        break;

                    case "CreateEvent": // #2
                        h.description.setText(Html.fromHtml("Created repository <b>" + e.getRepo().getName() + "</b>"));
                        h.linearLayout.setOnClickListener(RepoListener(s[0], s[1]));
                        break;

                    case "DeleteEvent": // #3   NEED TEST
                        DeletePayload deletePayload = (DeletePayload)e.getPayload();
                        h.description.setText(Html.fromHtml("Deleted <b>" + deletePayload.getRefType() + "</b>"));
                        h.linearLayout.setOnClickListener(RepoListener(s[0], s[1]));
                        break;

                    // #4
                    // #5

                    case "DownloadEvent":   // #6 NEED TEST
                        DownloadPayload downloadPayload = (DownloadPayload)e.getPayload();
                        h.description.setText(Html.fromHtml("Downloaded <b>" + downloadPayload.getDownload().getName() + "</b>"));
                        h.linearLayout.setOnClickListener(RepoListener(s[0], s[1]));
                        break;

                    case "FollowEvent": // #7   DEPRECATED
                        FollowPayload followPayload = (FollowPayload)e.getPayload();
                        h.description.setText(Html.fromHtml("Followed <b>" + followPayload.getTarget().getLogin() + "</b>"));
                        h.linearLayout.setOnClickListener(UserListener(followPayload.getTarget().getLogin()));
                        break;

                    case "ForkEvent":   // #8
                        ForkPayload p = (ForkPayload)e.getPayload();
                        h.description.setText(Html.fromHtml("Forked <b>" + e.getRepo().getName() + "</b> to <b>" + p.getForkee().getOwner().getLogin() + "/" + p.getForkee().getName() + "</b>"));
                        h.linearLayout.setOnClickListener(RepoListener(p.getForkee().getOwner().getLogin(), p.getForkee().getName()));
                        break;

                    // #9

                    case "GistEvent":   // #10  DEPRECATED
                        GistPayload gistPayload = (GistPayload)e.getPayload();
                        h.description.setText(Html.fromHtml(gistPayload.getAction() + "<b>" + gistPayload.getGist().getDescription() + "</b>"));
                        h.linearLayout.setOnClickListener(GistListener(gistPayload.getGist().getId()));
                        break;

                    case "GollumEvent": // #11  NEED TEST
                        GollumPayload gollumPayload = (GollumPayload)e.getPayload();
                        h.description.setText(Html.fromHtml("Updated wiki in <b>" + e.getRepo().getName() + "</b>"));
                        break;

                    case "WatchEvent":
                        h.description.setText(Html.fromHtml("Starred <b>" + e.getRepo().getName() + "</b>"));
                        h.linearLayout.setOnClickListener(RepoListener(s[0], s[1]));
                        break;

                }

            }

        }

    }

    @Override
    public int getItemCount() {
        return eventList.size();
    }
}

File: app/src/main/java/giuliolodi/gitnav/ui/events/EventAdapter.kt
/*
 * Copyright 2017 GLodi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package giuliolodi.gitnav.ui.events

import android.app.Activity
import android.support.v7.widget.RecyclerView
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import giuliolodi.gitnav.R
import giuliolodi.gitnav.ui.repository.RepoActivity
import giuliolodi.gitnav.ui.user.UserActivity
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.row_event.view.*
import org.eclipse.egit.github.core.event.*
import org.ocpsoft.prettytime.PrettyTime

/**
 * Created by giulio on 16/05/2017.
 */
class EventAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var mEventList: MutableList<Event?> = mutableListOf()
    private val mPrettyTime: PrettyTime = PrettyTime()
    private val onUserClick: PublishSubject<String> = PublishSubject.create()
    private val onImageClick: PublishSubject<String> = PublishSubject.create()

    fun getUserClicks(): Observable<String> {
        return onUserClick
    }

    fun getImageClicks(): Observable<String> {
        return onImageClick
    }

    class EventHolder(root: View) : RecyclerView.ViewHolder(root) {
        fun bind (event: Event, p: PrettyTime) = with (itemView) {
            row_event_name.text = event.actor.login
            row_event_date.text = p.format(event.createdAt)
            Picasso.with(context).load(event.actor.avatarUrl).resize(100,100).centerCrop().into(row_event_image)

            row_event_name.setOnClickListener {  }

            val s = arrayListOf(event.repo.name.split("/"))

            if (event.payload != null) {
                when (event.type) {
                    "CommitCommentEvent" -> { // #1
                        val commitCommentPayload: CommitCommentPayload = event.payload as CommitCommentPayload
                        row_event_description.text = Html.fromHtml("Commented <b>" + commitCommentPayload.comment.path.substringAfterLast('/') + "</b> in <font color='#326fba'>" + event.repo.name + "</font>: <b>"+ commitCommentPayload.comment.body + "</b>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "CreateEvent" -> { // #2
                        row_event_description.text = Html.fromHtml("Created <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "DeleteEvent" -> { // #3
                        val deletePayload: DeletePayload = event.payload as DeletePayload
                        row_event_description.text = Html.fromHtml("Deleted " + deletePayload.refType + " <b>" + deletePayload.ref + "</b> in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)                        }
                    }
                    "DeploymentEvent" -> { // #4 DEFAULT
                        row_event_description.text = Html.fromHtml("Deployed in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)                        }
                    }
                    "DeploymentStatus" -> { // #5 DEFAULT
                        row_event_description.text = Html.fromHtml("DeploymentStatus in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)                        }
                    }
                    "DownloadEvent" -> { // #6 NEED TEST
                        val downloadPayload: DownloadPayload = event.payload as DownloadPayload
                        row_event_description.text = Html.fromHtml("Downloaded <font color='#326fba'>" + downloadPayload.download.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)                        }
                    }
                    "FollowEvent" -> { // #7 DEPRECATED
                        val followPayload: FollowPayload = event.payload as FollowPayload
                        row_event_description.text = Html.fromHtml("Followed <font color='#326fba'>" + followPayload.target.login + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(UserActivity.getIntent(context).putExtra("username", followPayload.target.login))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "ForkEvent" -> { // #8
                        val forkPayload: ForkPayload = event.payload as ForkPayload
                        row_event_description.text = Html.fromHtml("Forked <font color='#326fba'>" + event.repo.name + "</font> to <font color='#326fba'>" + forkPayload.forkee.owner.login + "/" + forkPayload.forkee.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "ForkApplyEvent" -> { // #9
                        val forkApplyPayload: ForkApplyPayload = event.payload as ForkApplyPayload
                        row_event_description.text = Html.fromHtml("Fork applied <b>" + forkApplyPayload.head + "</b>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "GistEvent" -> { // #10
                        val gistPayload: GistPayload = event.payload as GistPayload
                        row_event_description.text = Html.fromHtml(gistPayload.action.substring(0,1).toUpperCase() + gistPayload.action.substring(1) + " <b>" + gistPayload.gist.description + "</b>")
                        row_event_ll.setOnClickListener {
                            // gist
                        }
                    }
                    "GollumEvent" -> { // #11 NEED TEST
                        row_event_description.text = Html.fromHtml("Updated wiki in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "IssueCommentEvent" -> { // #12
                        val issueCommentPayload: IssueCommentPayload = event.payload as IssueCommentPayload
                        row_event_description.text = Html.fromHtml(issueCommentPayload.action.substring(0,1).toUpperCase() + issueCommentPayload.action.substring(1) + " comment on issue <b>#" + issueCommentPayload.issue.number.toString() + "</b> in <font color='#326fba'>" + event.repo.name + "</font>: <b>" + issueCommentPayload.comment.body + "</b>")
                        row_event_ll.setOnClickListener {
                            // issue
                        }
                    }
                    "IssuesEvent" -> { // #13
                        val issuesPayload: IssuesPayload = event.payload as IssuesPayload
                        row_event_description.text = Html.fromHtml(issuesPayload.action.substring(0,1).toUpperCase() + issuesPayload.action.substring(1) + " issue <b>#" + issuesPayload.issue.number.toString() + "</b> in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            // issue
                        }
                    }
                    "LabelEvent" -> { // #14 DEFAULT
                        row_event_description.text = Html.fromHtml("Labeled <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "MemberEvent" -> { // #15
                        val memberPayload: MemberPayload = event.payload as MemberPayload
                        row_event_description.text = Html.fromHtml(memberPayload.action.substring(0,1).toUpperCase() + memberPayload.action.substring(1) + " <b>" + memberPayload.member.login + "</b> to <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "MembershipEvent" -> { // #16 DEFAULT
                        row_event_description.text = "Added or removed from team"
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "MilestoneEvent" -> { // #17 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed milestone in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "OrganizationEvent" -> { // #18 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed event in org <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "PageBuildEvent" -> { // #19 DEFAULT
                        row_event_description.text = Html.fromHtml("Built page in: <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "ProjectCardEvent" -> { // #20 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed project card event in: <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "ProjectColumnEvent" -> { // #21 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed column card event in: <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "Projectevent" -> { // #22 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed project event in: <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "PublicEvent" -> { // #23 DEFAULT
                        row_event_description.text = Html.fromHtml("Made <font color='#326fba'>" + event.repo.name + "</font> open source")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "PullRequestEvent" -> { // #24
                        val pullRequestPayload: PullRequestPayload = event.payload as PullRequestPayload
                        row_event_description.text = Html.fromHtml(pullRequestPayload.action.substring(0,1).toUpperCase() + pullRequestPayload.action.substring(1) + " pull request <b>#" + pullRequestPayload.number.toString() + "</b> in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "PullRequestReviewEvent" -> { // #25 NEED TEST
                        row_event_description.text = Html.fromHtml("Submitted a pull request review in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "PullRequestReviewCommentEvent" -> { // #26
                        val pullRequestReviewCommentPayload: PullRequestReviewCommentPayload = event.payload as PullRequestReviewCommentPayload
                        row_event_description.text = Html.fromHtml(pullRequestReviewCommentPayload.action.substring(0,1).toUpperCase() + pullRequestReviewCommentPayload.action.substring(1) + " pull request review comment <b>#" + pullRequestReviewCommentPayload.pullRequest.number.toString() + "</b> in <font color='#326fba'>" + event.repo.name + "</font>: <b>" + pullRequestReviewCommentPayload.comment + "</b>")
                    }
                    "PushEvent" -> { // #27
                        val pushPayload: PushPayload = event.payload as PushPayload
                        if (pushPayload.commits.size == 1)
                            row_event_description.text = Html.fromHtml("Pushed a commit to <b>" + pushPayload.ref + "</b> in <font color='#326fba'>" + event.repo.name + "</font>")
                        else
                            row_event_description.text = Html.fromHtml("Pushed " + pushPayload.commits.size.toString() + " commits to <b>" + pushPayload.ref + "</b> in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "ReleaseEvent" -> { // #28 NEED TEST
                        val releasePayload: ReleasePayload = event.payload as ReleasePayload
                        row_event_description.text = Html.fromHtml("Published <b>" + releasePayload.release.name + "</>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "RepositoryEvent" -> { // #29 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed a repository event in <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "StatusEvent" -> { // #30 DEFAULT
                        row_event_description.text = Html.fromHtml("Changed status of a git commit")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "TeamEvent" -> { // #31 DEFAULT
                        row_event_description.text = Html.fromHtml("Performed team event")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "TeamAddEvent" -> { // #32 NEED TEST
                        val teamAddPayload: TeamAddPayload = event.payload as TeamAddPayload
                        row_event_description.text = Html.fromHtml("Added <b>" + teamAddPayload.repo.name + "</b> to <b>" + teamAddPayload.team.name + "</b>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }
                    "WatchEvent" -> { // #33
                        row_event_description.text = Html.fromHtml("Starred <font color='#326fba'>" + event.repo.name + "</font>")
                        row_event_ll.setOnClickListener {
                            context.startActivity(RepoActivity.getIntent(context).putExtra("owner", s[0][0]).putExtra("name", s[0][1]))
                            (context as Activity).overridePendingTransition(0,0)
                        }
                    }

                }
            }
        }
    }

    class LoadingHolder(root: View) : RecyclerView.ViewHolder(root)

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {
        val root: RecyclerView.ViewHolder
        if (viewType == 1) {
            val view = (LayoutInflater.from(parent?.context).inflate(R.layout.row_event, parent, false))
            root = EventHolder(view)
        } else  {
            val view = (LayoutInflater.from(parent?.context).inflate(R.layout.row_loading, parent, false))
            root = LoadingHolder(view)
        }
        return root
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is EventHolder) {
            val event = mEventList[position]!!
            holder.bind(event, mPrettyTime)
            holder.itemView.row_event_image.setOnClickListener { event.actor?.login?.let { onImageClick.onNext(it) } }
            holder.itemView.row_event_name.setOnClickListener { event.actor?.login?.let { onImageClick.onNext(it) } }
        }
    }

    override fun getItemViewType(position: Int): Int { return if (mEventList[position] != null) 1 else 0 }

    override fun getItemCount(): Int { return mEventList.size }

    override fun getItemId(position: Int): Long { return position.toLong() }

    fun addEvents(eventList: List<Event>) {
        if (!eventList.isEmpty()) {
            val lastItemIndex = if (mEventList.size > 0) mEventList.size else 0
            mEventList.addAll(eventList)
            notifyItemRangeInserted(lastItemIndex, mEventList.size - 1)
        }
        /*
        if (mEventList.isEmpty()) {
            mEventList.clear()
            mEventList.addAll(eventList)
            notifyDataSetChanged()
        }
        else if (mEventList.lastIndexOf(null) != -1) {
            val lastNull = mEventList.lastIndexOf(null)
            mEventList.removeAt(lastNull)
            notifyItemRemoved(lastNull)
            mEventList.addAll(eventList)
            notifyItemRangeInserted(lastNull, eventList.size - 1)
        }
        else {
            val lastItemIndex = mEventList.size - 1
            mEventList.addAll(eventList)
            notifyItemRangeInserted(lastItemIndex, mEventList.size - 1)
        }
        */
    }

    fun showLoading() {
        mEventList.add(null)
        notifyItemInserted(mEventList.size - 1)
    }

    fun hideLoading() {
        val lastNull = mEventList.lastIndexOf(null)
        if (lastNull != -1) {
            mEventList.removeAt(lastNull)
            notifyItemRemoved(lastNull)
        }
    }

    fun clear() {
        mEventList.clear()
        notifyDataSetChanged()
    }

}
